### PR TITLE
custom controller for youtube-dl

### PR DESCRIPTION
**Pull request by samsullivan**

### CODE REVIEW

File: packages/@uppy/companion/src/server/controllers/youtube.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    getSize: () => {
+      const { size } = statSync(tmpPath)
+      return size
+    },
+    download: () => createReadStream(tmpPath),
+    onUnhandledError: err => {
+      logger.error(err, 'controller.youtube.upload.error', req.id)
+      res.status(500).json({ message: 'Failed to upload video' })
+    },
+  })
```
Comment (samsullivan) [inline packages/@uppy/companion/src/server/controllers/youtube.js]: There's no great (easy) way to clean up this file afterwards. I could probably make it work with some extra effort, but I noticed you were temporarily not unlinking the temporary yt-dl stuff already -- maybe this is good enough for now?

File: packages/@uppy/companion/src/companion.js [hunk 2]
```diff
@@ -102,6 +102,11 @@ module.exports.app = (options = {}) => {
   app.use('/s3', s3(options.providerOptions.s3))
   app.use('/url', url())
 
+  // hooks directly into youtube-dl while avoiding Uppy's meta endpoints
+  // previously, we tried to implement this w/in url controller but had tough to debug problems
+  app.post('/youtube/download', controllers.youtube.download)
+  app.post('/youtube/upload', controllers.youtube.upload)
```
Comment (samsullivan) [inline packages/@uppy/companion/src/companion.js:108]: DL/UL routes are separate mainly for better UX. Currently, to upload via tusd, we're using `uppy.addFile()`. This closes the plugin UI on submission and shows errors in a toast (compared to showing an error message in the plugin UI). Also, for longer videos, you actually see the filesize progress upload after the initial download is finished.

We're already needing to download the file completely before using tusd, so this isn't too much different.

File: packages/@uppy/companion/src/server/controllers/youtube.js (hunk 1 above)
Comment (zackbloom) [reply packages/@uppy/companion/src/server/controllers/youtube.js]: I'm sure it's ok for now, yea

Comment (samsullivan) [reply packages/@uppy/companion/src/server/controllers/youtube.js]: Ended up improving this in https://github.com/SendHeirloom/uppy/pull/3.

**Pull request merged**